pkg/operator/staticpod/controller/staticpodstate/staticpodstate_controller.go

```diff
@@ -0,0 +1,166 @@
+package staticpodstate
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+var (
+	staticPodStateControllerFailing      = "StaticPodsFailing"
+	staticPodStateControllerWorkQueueKey = "key"
+)
+
+// StaticPodStateController is a controller that watches the currentRevision and targetRevision fields for each node and spawn
+// installer pods to update the static pods on the master nodes.
+type StaticPodStateController struct {
+	targetNamespace, staticPodName string
+
+	operatorConfigClient v1helpers.StaticPodOperatorClient
+	kubeClient           kubernetes.Interface
+	eventRecorder        events.Recorder
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+}
+
+// NewStaticPodStateController creates a controller that watches static pods and will produce a failing status if the
+// static pods start crashing for some reason.
+func NewStaticPodStateController(
+	targetNamespace, staticPodName string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	operatorConfigClient v1helpers.StaticPodOperatorClient,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) *StaticPodStateController {
+	c := &StaticPodStateController{
+		targetNamespace: targetNamespace,
+		staticPodName:   staticPodName,
+
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+		eventRecorder:        eventRecorder,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StaticPodStateController"),
+	}
+
+	operatorConfigClient.Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForTargetNamespace.Core().V1().Pods().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *StaticPodStateController) sync() error {
+	operatorSpec, originalOperatorStatus, _, err := c.operatorConfigClient.GetStaticPodOperatorState()
+	if err != nil {
+		return err
+	}
+
+	switch operatorSpec.ManagementState {
+	case operatorv1.Unmanaged:
+		return nil
+	case operatorv1.Removed:
+		// TODO probably just fail.  Static pod managers can't be removed.
+		return nil
+	}
+
+	errs := []error{}
```

Comment:
hrm, we still don't have good lib to deal with error aggregation?

Reply:
There's https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go, but that's useful only when you want to filter through them. What's wrong with an array?